s3transfer/upload.py

```diff
@@ -24,6 +24,49 @@
 from s3transfer.utils import DeferredOpenFile
 
 
+class InterruptReader(object):
+    """Wrapper that can interrupt reading using an error
+
+    It uses a transfer coordinator to propogate an error if it notices
+    that a read is being made while the file is being read from.
+
+    :type fileobj: file-like obj
+    :param fileobj: The file-like object to read from
+
+    :type transfer_coordinator: s3transfer.futures.TransferCoordinator
+    :param transfer_coordinator: The transfer coordinator to use if the
+        reader needs to be interrupted.
+    """
+    def __init__(self, fileobj, transfer_coordinator):
+        self._fileobj = fileobj
+        self._transfer_coordinator = transfer_coordinator
+
+    def read(self, amount=None):
+        # If there is an exception, then raise the exception.
+        # We raise an error instead of returning no bytes because for
+        # requests where the content length and md5 was sent, it will
+        # cause md5 mismatches and retries as there was no indication that
+        # the stream being read from encountered any issues.
+        if self._transfer_coordinator.exception:
+            raise self._transfer_coordinator.exception
```

Comment:
This is going to acquire a shared lock twice.  I would imagine this could slow things down as you ramp up your parallelism.

I don't think `self._exception` needs to be protected by a lock does it?  Reading from it should be thread safe.  What do you think about removing the lock?